File: src/main/java/todayquest/quest/dto/QuestResponseDto.java
package todayquest.quest.dto;

import lombok.Builder;
import lombok.Data;
import org.springframework.format.annotation.DateTimeFormat;
import todayquest.quest.entity.Quest;
import todayquest.quest.entity.QuestDifficulty;
import todayquest.quest.entity.QuestState;
import todayquest.quest.entity.QuestType;
import todayquest.reward.dto.RewardResponseDto;

import java.time.LocalDate;
import java.time.LocalTime;
import java.util.List;
import java.util.stream.Collectors;

@Builder
@Data
public class QuestResponseDto {

    private Long questId;
    private String title;
    private String description;
    private boolean isRepeat;
    @DateTimeFormat(pattern = "yyyy-MM-dd")
    private LocalDate deadLineDate;
    @DateTimeFormat(pattern = "HH:mm")
    private LocalTime deadLineTime;
    private QuestState state;
    private QuestType type;
    private QuestDifficulty difficulty;
    private List<RewardResponseDto> rewards;

    public static QuestResponseDto createDto(Quest quest) {
        return QuestResponseDto.builder()
                .questId(quest.getId())
                .title(quest.getTitle())
                .description(quest.getDescription())
                .isRepeat(quest.isRepeat())
                .deadLineDate(quest.getDeadLineDate())
                .deadLineTime(quest.getDeadLineTime())
                .state(quest.getState())
                .type(quest.getType())
                .difficulty(quest.getDifficulty())
                .rewards(quest.getRewards().stream().map(r -> RewardResponseDto.createDto(r.getReward())).collect(Collectors.toList()))
                .build();
    }

}

File: src/main/java/todayquest/quest/dto/QuestResponseDto.kt
package todayquest.quest.dto

import org.springframework.format.annotation.DateTimeFormat
import todayquest.quest.entity.DetailQuestState
import todayquest.quest.entity.Quest
import todayquest.quest.entity.QuestDifficulty
import todayquest.quest.entity.QuestState
import todayquest.reward.dto.RewardResponseDto
import java.time.LocalDate
import java.time.LocalDateTime
import java.time.LocalTime

data class QuestResponseDto(
    var questId: Long? = null,
    var title: String? = null,
    var description: String? = null,
    var seq: Long? = null,
    @field:DateTimeFormat(pattern = "yyyy-MM-dd")
    var deadLineDate: LocalDate? = null,
    @field:DateTimeFormat(pattern = "HH:mm")
    var deadLineTime: LocalTime? = null,
    var state: QuestState? = null,
    var difficulty: QuestDifficulty? = null,
    var rewards: List<RewardResponseDto>? = null,
    var lastModifiedDate: LocalDateTime? = null,
    var detailQuests: List<DetailQuestResponseDto>? = null,
    var canComplete : Boolean? = null,
) {

    companion object {
        @JvmStatic
        fun createDto(quest: Quest): QuestResponseDto {
            return QuestResponseDto(
                questId = quest.id,
                title = quest.title,
                description = quest.description,
                seq = quest.seq,
                deadLineDate = quest.deadLineDate,
                deadLineTime = quest.deadLineTime,
                state = quest.state,
                difficulty = quest.difficulty,
                lastModifiedDate = quest.lastModifiedDate,
                rewards = quest.rewards.map {
                    RewardResponseDto.createDto(
                        it.reward
                    )
                }.toCollection(mutableListOf()),
                detailQuests = quest.detailQuests.map {
                    DetailQuestResponseDto.createDto(
                        it
                    )
                }.toCollection(mutableListOf()),
                canComplete = quest.detailQuests.all { dq -> dq.state == DetailQuestState.COMPLETE }
            )
        }
    }

}